sort handles incomparable values

Akirathan opened the pull request:
Fixes #5742 

### Pull Request Description
`Vector.sort` handles incomparable values. Now, it should sort basically any vector, potentially attaching a warning if `sort` encountered incomparable values.

### Important Notes
- Slight change of `Vector.sort` method signature
  - To be able to introduce `Vector.sort_builtin` optimized builtin for the most common case - sorting only primitive values with `Default_Comparator`.

### Checklist

Please ensure that the following checklist has been satisfied before submitting the PR:

- [x] The documentation has been updated, if necessary.
- [x] Screenshots/screencasts have been attached, if there are any visual changes. For interactive or animated visual changes, a screencast is preferred.
- [x] All code follows the
      [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md),
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
      and
      [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md)
      style guides. In case you are using a language not listed above, follow the [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md) style guide.
- All code has been tested:
  - [x] Unit tests have been written where possible.
  - [x] If GUI codebase was changed, the GUI was tested when built using `./run ide build`.

Commit:
Remove the remaining comparison operator overrides for numbers.

Commit:
Update type ascriptions in some operators in Any

Commit:
Add @GenerateUncached to AnyToTextNode.

Will be used in another node with @GenerateUncached.

Commit:
Add tests for "sort handles incomparable types"

Commit:
Vector.sort handles incomparable types

JaroslavTulach commented on test/Tests/src/Data/Ordering_Spec.enso:
Good set of tests at first sight.

Commit:
Implement sort handling for different comparators

Commit:
Comparison operators in Any do not throw Type_Error

Commit:
Fix some issues in Ordering_Spec

Commit:
Consolidate all sorting functionality into a single builtin node.

Commit:
Fix warnings attachment in sort

Commit:
PrimitiveValuesComparator handles other types than primitives

Commit:
Merge branch 'develop' into wip/akirathan/sort-incomparable-5742

# Conflicts:
#	distribution/lib/Standard/Base/0.0.0-dev/src/Any.enso
#	distribution/lib/Standard/Base/0.0.0-dev/src/Data/Ordering.enso
#	test/Tests/src/Data/Numbers_Spec.enso
#	test/Tests/src/Data/Ordering_Spec.enso

Commit:
Fix byFunc calling

Commit:
on function can be called from the builtin

Commit:
Merge branch 'develop' into wip/akirathan/sort-incomparable-5742

Commit:
Merge branch 'develop' into wip/akirathan/sort-incomparable-5742

Commit:
Fix build of native image

Commit:
Update changelog

Commit:
Add VectorSortTest

Commit:
Merge branch 'develop' into wip/akirathan/sort-incomparable-5742

Commit:
Builtin method should not throw DataflowError.

If yes, the message is discarded (a bug?)

Commit:
TypeOfNode may not return only Type

Commit:
UnresolvedSymbol is not supported as `on` argument to Vector.sort_builtin

Commit:
Fix docs

Commit:
Fix bigint spec in LessThanNode

Commit:
Small fixes

Commit:
Small fixes

Commit:
Nothings and Nans are sorted at the end of default comparator group.

But not at the whole end of the resulting vector.

Akirathan commented on test/Table_Tests/src/In_Memory/Table_Spec.enso:
```diff
@@ -472,7 +472,7 @@ spec =
             cmp a b = Ordering.compare (a.x-a.y).abs (b.x-b.y).abs
             r = c.sort by=cmp
-            r.to_vector.should_equal [My.Data 1 2, My.Data 3 4, My.Data 1 0, My.Data 2 5, My.Data 6 3, Nothing]
+            r.to_vector.should_equal [Nothing, My.Data 1 2, My.Data 3 4, My.Data 1 0, My.Data 2 5, My.Data 6 3]
```
This is not according to the specs that we have in the issue description - Nothing should be at the very end of the sorted vector. This might be a bit tricky for me to implement now. Currently, all the value with the default comparator (both primitive and non-primitive) are sorted at the beginning of the resulting vector, with Nothing and Nans placed at the end of the group.

So the resulting vector has these sections:
```
[<primitive values>, <non-primitive values with Default_Comparator>, <incomparable values with Default_Comparator (NaN, Nothing)>, <other values with different comparators - sorted by FQN of comparators>]
```

TL;DR: Should I revert this change and try harder to implement it, or are we OK with this so far?

cc @JDunkerley

Akirathan replied:
cc @radeusgd

radeusgd replied:
I guess it's up-to @jdunkerley , as he was the one wanting to have `Nothing` at the end. If this is the requirement, I think we should do it indeed.

I'm fine with doing a fix-up in a separate PR if you think it's better - but would be good to create a ticket for this in that case, as otherwise this will get lost.

IMO though the best way is to fix it now, with some simple workaround. I imagine we could just have a completely special-case branch for `interop.isNull(x)` and put these at the end. Just do a check for null _before_ check for comparator and treat nulls as a _separate_ group from the `DefaultComparator` group. Would that work?

Commit:
Fix checking of `by` parameter - now accepts functions with default arguments.

JaroslavTulach commented:
I see just two failures:
- [DebuggingEnsoTest](https://github.com/enso-org/enso/actions/runs/4621375634/jobs/8172721591?pr=5998#step:10:4749) probably misses an import(!?)
- [RuntimeAsyncCommandTest](https://github.com/enso-org/enso/actions/runs/4621375634/jobs/8172721591?pr=5998#step:10:5033) - dtto

Looks promising and I can't wait to see this PR integrated.

Commit:
Fix changelog formatting

Commit:
Fix imports in DebuggingEnsoTest

Commit:
Remove Array.sort_builtin

Commit:
Merge branch 'develop' into wip/akirathan/sort-incomparable-5742

# Conflicts:
#	CHANGELOG.md

Commit:
Add comparison operators to micro-distribution

hubertp commented on engine/runtime/src/main/java/org/enso/interpreter/node/expression/builtin/ordering/SortVectorNode.java:
```diff
@@ -0,0 +1,725 @@
+    }
+  }
+
+  private boolean isBuiltinType(Object type) {
```
is that necessary? Types have `isBuiltin` method

Akirathan replied:
I am afraid so. This method should be called rather something like `isBuiltinTypeAndCanBeCompared`. I want to be able to explicitly control the list of builtin types for which I know that it makes sense to compare them.

hubertp commented on engine/runtime/src/main/java/org/enso/interpreter/node/expression/builtin/ordering/SortVectorNode.java:
```diff
@@ -0,0 +1,725 @@
+    return isNothing(object, EnsoContext.get(this));
+  }
+
+  private boolean isNothing(Object object, EnsoContext ctx) {
```
There is `isNothing` method on Type and `isNull` on `InteropLibrary` already

Akirathan replied:
I don't want `InteropLibrary.isNull` here, I want strictly `Nothing` type. Moreover, I don't have `Type` but `Object`.

radeusgd replied:
Why is that? Shouldn't our `Nothing` be basically indistinguishable from `null` from other languages?

Akirathan replied:
Probably yes. But I want to just recognize our `Nothing`, not nulls from other languages, there is definitely no interop here. So we can omit the `InteropLibrary` abstraction.

radeusgd replied:
I'm not so sure. Of course this is an extremely rare case but let's imagine I have:
```
my_on = special_function_that_determines_what_on_to_use_for_sorting
vec.sort on=my_on
```

The user is free to set `on` to whatever. Now imagine that `special_function_that_determines_what_on_to_use_for_sorting` calls some interop JS method to return a custom function or `null` to use the default. Bam! Got a likely interop null, I'm not quite sure if it's converted automatically to a `Nothing` - I doubt it.

I know it's a rare edge case - but this is how hard to debug bugs are born. Someday someone will use interop to generate `on` and will get bitten. And what benefit is there to _not using the preferred way for a null-check_ and instead use something non-standard?

So I suggest to use `InteropLibrary` which gives us confidence that this works in a standard way and we don't have to worry about it. And if not, I insist to add a check if a `null` returned from JS will yield the correct behaviour in the current state of things. If it does, OK, I guess why not keep it. But IMO it's just much clearer to use the standard means.

Akirathan replied:
I guess that you are right, done in cc1f5274345004f8d4518812766cb9e270c68b8f

hubertp commented on engine/runtime/src/main/java/org/enso/interpreter/node/expression/builtin/ordering/SortVectorNode.java:
```diff
@@ -0,0 +1,725 @@
+        // This case is not supported yet, as it requires topological sorting.
+        // We cannot detect if the result was actually returned from the default comparator (it
+        // could have been transitively called), so we just bailout.
+        throw new CompareException(x, y);
```
do we have a test for this case?

Akirathan replied:
Yes, I have added the test case in b39ea29d5b7f8d4cab83a79cb932fa6042e1df3c

hubertp commented on engine/runtime/src/main/java/org/enso/interpreter/node/expression/builtin/ordering/SortVectorNode.java:
```diff
@@ -0,0 +1,725 @@
+
+    private boolean hasFunctionSelfArgument(Function function) {
+      if (function.getSchema().getArgumentsCount() > 0) {
+        return function.getSchema().getArgumentInfos()[0].getName().equals("self");
```
1) it's not guaranteed that the first argument is self 2. I believe we have synthetic `self` always, even if not specified explicitly, so pretty much every function has that?

Akirathan replied:
I have observed that for instance and static methods, the first argument is always `self` (for static methods it is synthetic), and for, e.g., lambda functions, the first argument is not self. This is important in order to call the method with correct number of arguments, we want to support calling both lambda functions and instance/static methods.

hubertp commented on engine/runtime/src/main/java/org/enso/interpreter/runtime/util/Collections.java:
```diff
@@ -0,0 +1,78 @@
+    }
+
+    public ArrayListObj() {
+      this(8);
```
And the magic `8` stems from?

Akirathan replied:
Not really. Besides its not used anywhere, so I am going to remove it anyway.

hubertp reviewed (approved):
mostly nits, looks good overall

Commit:
Remove Array.sort_builtin

Commit:
Replace Incomparable_Values by Type_Error in some tests

Commit:
Add on_incomparable argument to Vector.sort_builtin

Commit:
Merge branch 'develop' into wip/akirathan/sort-incomparable-5742

# Conflicts:
#	CHANGELOG.md
#	distribution/lib/Standard/Base/0.0.0-dev/src/Data/Vector.enso
#	test/Tests/src/Data/Vector_Spec.enso

Commit:
Fix after merge - Array.sort delegates to Vector.sort

JaroslavTulach commented:
Good, there is just [VectorArraySignatureTest failure](https://github.com/enso-org/enso/actions/runs/4688223391/jobs/8308469437?pr=5998#step:10:4674) and that shall be easy to fix. It checks if the documentation for `Vector` and `Array` functions is identical (with few transformations).

Commit:
Add more tests for problem_behavior on Vector.sort

Commit:
SortVectorNode throws only Incomparable_Values.

Commit:
Delete Collections helper class

Commit:
Add test for expected failure for custom incomparable values

Akirathan commented on distribution/lib/Standard/Base/0.0.0-dev/src/Data/Index_Sub_Range.enso:
```diff
@@ -117,7 +117,7 @@ invert_range_selection ranges length needs_sorting =
 sort_and_merge_ranges : Vector Range -> Vector Range
 sort_and_merge_ranges ranges =
-    sorted = ranges.filter (range-> range.is_empty.not) . sort on=(.start)
+    sorted = ranges.filter (range-> range.is_empty.not) . sort on=(_.start)
```
`(.start)` is parsed to `UnresolvedSymbol<start>`, and I found no easy way how to call that. Whereas `(_.start)` is parsed into `Function`. I created issue #6726 to track that.

radeusgd replied:
#6276 *

Commit:
Cosmetics.

Commit:
Fix test expecting different comparators warning

Commit:
isNothing is checked via interop

Akirathan commented:
There is `Test org.enso.compiler.SerdeCompilerTest.testFibTest failed: org.graalvm.polyglot.PolyglotException: java.util.ConcurrentModificationException` error in a Windows build: https://github.com/enso-org/enso/actions/runs/4697459629/jobs/8329091090?pr=5998#step:10:4659. Seems like a transient? Or can it be caused by me introducing more loggers that are initialized with something like `TruffleLogger.getLogger("enso", SortVectorNode.class);`, for example [here](https://github.com/enso-org/enso/pull/5998/files#diff-481daf10d9ab1323a1cbec24cf7c449f24844247e0c05917f13b8f39f0915b49R61)? Restarting the jobs now. Note that linux job succeeded.

<blockquote><img src="https://opengraph.githubassets.com/6cf40e2b01a459b39eb61c641c3436d2791fe63511fb25ddbccd7051acb83a1a/enso-org/enso" width="48" align="right"><div><img src="https://github.githubassets.com/favicons/favicon.svg" height="14"> GitHub</div><div><strong><a href="https://github.com/enso-org/enso">sort handles incomparable values · enso-org/enso@cc1f527</a></strong></div><div>Hybrid visual and textual functional programming. Contribute to enso-org/enso development by creating an account on GitHub.</div></blockquote>

Akirathan commented:
Seems like the aforementioned failure in `SerdeCompilerTest` is not a transient one. Encountered it another time in https://github.com/enso-org/enso/actions/runs/4697459629/jobs/8330446700?pr=5998#step:10:4665. I will refactor the newly introduced logger in `SortVectorNode` such that it creates the logger via `EnsoContext.getLogger`, rather than via `TruffleLogger.get`.

<blockquote><img src="https://opengraph.githubassets.com/6cf40e2b01a459b39eb61c641c3436d2791fe63511fb25ddbccd7051acb83a1a/enso-org/enso" width="48" align="right"><div><img src="https://github.githubassets.com/favicons/favicon.svg" height="14"> GitHub</div><div><strong><a href="https://github.com/enso-org/enso">sort handles incomparable values · enso-org/enso@cc1f527</a></strong></div><div>Hybrid visual and textual functional programming. Contribute to enso-org/enso development by creating an account on GitHub.</div></blockquote>

radeusgd commented on distribution/lib/Standard/Base/0.0.0-dev/src/Data/Array.enso:
```diff
@@ -160,13 +161,17 @@ type Array
        - order: The order in which the array elements are sorted.
        - on: A projection from the element type to the value of that element
-         being sorted on.
+         being sorted on. If set to `Nothing` (the default argument),
```
```suggestion
         being sorted on. If set to `Nothing` (the default),
```

IMO this reads better that way.

radeusgd commented on distribution/lib/Standard/Base/0.0.0-dev/src/Data/Array.enso:
```diff
@@ -197,8 +214,17 @@ type Array
-    sort self (order = Sort_Direction.Ascending) (on = x -> x) (by = (Ordering.compare _ _)) = Vector.sort self order on by
+
+       > Example
+         Sorting an array with elements with different comparators. Values 1
+         and My_Type have different comparators. 1 will be sorted before My_Type
```
```suggestion
         Sorting an array with elements with different comparators. Values `1`
         and `My_Type` have different comparators. `1` will be sorted before `My_Type`
```

radeusgd commented on distribution/lib/Standard/Base/0.0.0-dev/src/Data/Array.enso:
```diff
@@ -160,13 +161,17 @@ type Array
+         elements, returning an an `Ordering` if the two elements are comparable
+         or `Nothing` if they are not. If set to `Nothing` (the default argument),
+         `Ordering.compare _ _` method will be used.
+       - on_incomparable: A `Problem_Behavior` specifying what should happen if
+         two incomparable values are encountered.
```
What happens if values having _the same_ comparator still are not comparable? I assume the warning is still issued, I guess the order is then just undefined? I would like this case to be documented too.

(but btw. very great docs for the _'? Multiple comparators'_ case, kudos)

Akirathan replied:
Added additional doc section for that

radeusgd commented on distribution/lib/Standard/Base/0.0.0-dev/src/Data/Numbers.enso:
```diff
@@ -288,7 +289,7 @@ type Number
          Check if 1 is equal to 1.0000001 within 0.001.
 
              1.equals 1.0000001 epsilon=0.001
-    equals : Number -> Number -> Boolean
+    equals : Number -> Number -> Boolean ! Incomparable_Values
     equals self that epsilon=0.0 =
         (self == that) || ((self - that).abs <= epsilon)
```
Are you sure about this signature?

When can this yield `Incomparable_Values`? I assume only Numbers will go in here.

I guess with `NaN`? Is `NaN` the **only** possibility of getting `Incomparable_Values` or are other situations that I did not predict?

I'd appreciate a comment explaining the `! Error Conditions` here.

Akirathan replied:
Yes. It throws `Incomparable_Values` error only iff NaN is one of the given parameters. I updated the docs.

radeusgd commented on distribution/lib/Standard/Base/0.0.0-dev/src/Data/Vector.enso:
```diff
@@ -17,12 +16,20 @@ import project.Errors.Common.Not_Found
 
+import project.IO
+
+# We have to import also conversion methods, therefore, we import all from the Ordering
+# module
```
```suggestion
## We have to import also conversion methods, therefore, we import all from the Ordering
   module
```

I think the convention is to use `##` for the multiline comments. Unless there was some issue with it?

radeusgd replied:
Btw. I think this raises a good point that we may want to have a way to import conversion methods _without_ necessarily importing `all`. What do you think? Shall we start a discussion?

I guess one way to do so should be to do `import project.Data.Ordering` instead. I'm pretty sure the not-`from` imports _do_ import the extension methods within.

radeusgd replied:
But maybe a more explicit way for this would be good too. Just a random thought, low priority. Definitely independent of this PR.

Akirathan replied:
> Shall we start a discussion?

I don't know. I guess it depends on whether the current behavior is the expected one, i.e., if I import a type with `import Standard.Base.Data.Ordering.Comparable.Comparable`, then all its extension methods from that module are imported as well, but not any conversion methods - neither those with `Comparable` as target, neither those with `Comparable` as source. Is this intuitive/expected/bug?

radeusgd replied:
Honestly, I don't know.

My intuition would be that if I import a type, probably it's extension methods are imported too... But probably only its and not others.

I'd appreciate a way to import _only_ extension methods from a file, but that's just my subjective view. Also definitely not something of high priority, as we are doing fine with current state of the import system, I think.

radeusgd commented on distribution/lib/Standard/Base/0.0.0-dev/src/Data/Vector.enso:
```diff
@@ -878,16 +901,25 @@ type Vector a
+            True -> self.map it-> Comparable.from it
+            False -> self.map it-> Comparable.from (on it)
+        compare_funcs = comps.map (it-> it.compare)
+        problem_behavior = case on_incomparable of
+            Problem_Behavior.Ignore -> 0
+            Problem_Behavior.Report_Warning -> 1
+            Problem_Behavior.Report_Error -> 2
```
Ugh. This looks very meh. Although indeed I'm not sure what better way there is to do this without making `Problem_Behavior` a builtin. Maybe we just should make this Atom a builtin if it's used here??

If not, I'd at least move this mapping to some external method, probably in `Problem_Behavior.enso` as a helper, _not_ on the Problem_Behavior type but on module to keep it hidden. Ideally I'd use some constants coming from Java for these numbers. Maybe you can define an `enum` in the core polyglot utils to access it from both sides. Although this starts sounding like an overkill for sth so simple, so up to you.

But I'd at least split it off to a separate method - if we ever need this mapping in another place, I don't want it being copied-and-pasted, let's have a single method to reference.

Akirathan replied:
Refactored this into a private, documented, method `Problem_Behavior.to_number`.

radeusgd commented on distribution/lib/Standard/Test/0.0.0-dev/src/Problems.enso:
```diff
@@ -3,7 +3,7 @@ from Standard.Base import all
 from project import Test
 import project.Extensions
 
-## Returns values of warnings attached to the value.Nothing
```
Whooops, I guess that's me and my weird IntelliSense settings. Sorry. And thanks for noticing and fixing 🙂

radeusgd commented on engine/runtime/src/main/java/org/enso/interpreter/node/expression/builtin/ordering/SortArrayNode.java:
```diff
@@ -0,0 +1,32 @@
+import org.enso.interpreter.runtime.state.State;
+
+/**
+ * Just a wrapper for {@code Array.sort_builtin} that delegates to {@code Vector.sort_builtin}. This
+ * is a temporary workaround until Array and Vector are fully merged into a single type.
```
```suggestion
 * Just a wrapper for {@code Array.sort_builtin} that delegates to {@code Vector.sort_builtin}.
```

Is the merger planned? From what I remember from [our discussions with @JaroslavTulach](https://discord.com/channels/@me/948519891961544714/1095320137873227826) we decided that they actually do have slightly different meanings, so we probably want to keep them distinct.

radeusgd replied:
Btw. do we need this? Cannot we just make `Array.sort` delegate to `Vector` somehow and have just one builtin method to do this??

Akirathan replied:
We cannot, because builtin methods cannot be called as static methods - #6282. I have tried that already.

radeusgd commented on test/Tests/src/Data/Ordering_Spec.enso:
```diff
@@ -36,9 +47,33 @@ Comparable.from (_ : UPair) = UPair_Comparator
+expect_incomparable_warn : Any -> Any -> Any -> Nothing
+expect_incomparable_warn left_val right_val result =
+    # Incomparable values warning wraps Text values in simple quotes
+    left_val_text = if Meta.is_a left_val Text then "'" + left_val + "'" else left_val.to_text
+    right_val_text = if Meta.is_a right_val Text then "'" + right_val + "'" else right_val.to_text
```
```suggestion
    left_val_text = left_val.pretty
    right_val_text = right_val.pretty
```

Won't this work?

Akirathan replied:
It should, thanks for the suggestion.

radeusgd commented on test/Tests/src/Data/Ordering_Spec.enso:
```diff
@@ -36,9 +47,33 @@ Comparable.from (_ : UPair) = UPair_Comparator
+
+# === The Tests ===
 spec =
+    topo_sort_pending = "Waiting for implementation of topological sort (https://github.com/enso-org/enso/issues/5742)"
```
```suggestion
    topo_sort_pending = "Waiting for implementation of topological sort (https://github.com/enso-org/enso/issues/5834)"
```

Shouldn't we link to the [topo-sort issue](https://github.com/enso-org/enso/issues/5834) instead of [the current one, attached to this PR](https://github.com/enso-org/enso/issues/5742)?

Akirathan replied:
Definitely. I made a type. thx.

radeusgd commented on test/Tests/src/Data/Ordering_Spec.enso:
```diff
@@ -93,10 +128,106 @@ spec =
+            [My_Type.Value 1, Ord.Value 1, 20, 10].sort . should_equal [10, 20, My_Type.Value 1, Ord.Value 1]
+            [Ord.Value 1, 20, 10, My_Type.Value 1].sort . should_equal [10, 20, My_Type.Value 1, Ord.Value 1]
+
+        Test.specify "should be able to sort even unordered values" pending="https://github.com/enso-org/enso/issues/5834" <|
```
```suggestion
        Test.specify "should be able to sort even unordered values" pending=topo_sort_pending  <|
```

radeusgd commented on test/Tests/src/Data/Ordering_Spec.enso:
```diff
@@ -93,10 +128,106 @@ spec =
+            [Ord.Value 2, UPair.Value "a" "b", Ord.Value 1, UPair.Value "c" "d"].sort . should_equal [Ord.Value 2, Ord.Value 1, UPair.Value "a" "b", UPair.Value "c" "d"]
+            [Ord.Value 2, UPair.Value "X" "Y", Ord.Value 1, UPair.Value "c" "d"].sort . should_equal [Ord.Value 2, Ord.Value 1, UPair.Value "X" "Y", UPair.Value "c" "d"]
+
+        Test.specify "should produce warning when sorting unordered values" pending="https://github.com/enso-org/enso/issues/5834" <|
```
```suggestion
        Test.specify "should produce warning when sorting unordered values" pending=topo_sort_pending  <|
```

radeusgd commented on test/Tests/src/Data/Ordering_Spec.enso:
```diff
@@ -93,10 +128,106 @@ spec =
+            expect_incomparable_warn Nothing Number.nan <| [Nothing, Number.nan].sort on_incomparable=Problem_Behavior.Report_Warning
+            expect_incomparable_warn 1 "hello" <| [1, "hello"].sort on_incomparable=Problem_Behavior.Report_Warning
+
+        Test.specify "should respect previous warnings on a vector" <|
+            Problems.expect_warning "my_warn" <| (Warning.attach "my_warn" [3, 2]) . sort
+            Problems.expect_warning "my_warn" <| (Warning.attach "my_warn" [3, Number.nan]) . sort
+            expect_incomparable_warn 3 Number.nan <| (Warning.attach "my_warn" [3, Number.nan]) . sort on_incomparable=Problem_Behavior.Report_Warning
+
+        Test.specify "should respect previous warnings on vectors" pending="https://github.com/enso-org/enso/issues/6070" <|
+            Problems.expect_warning "my_warn" <| [3, Warning.attach "my_warn" 2].sort
+            expect_incomparable_warn 1 Number.nan [1, Warning.attach "my_warn" Number.nan].sort
+            Problems.expect_warning "my_warn" <| [1, Warning.attach "my_warn" Number.nan].sort
```
I appreciate.

I wanted to ask if we can have tests with warnings-attached-values, since they've been giving us some pain recently. But I see these are already there. Awesome!

radeusgd commented on test/micro-distribution/lib/Standard/Base/0.0.0-dev/src/Any.enso:
```diff
@@ -3,8 +3,12 @@ type Any
     == self other = Comparable.equals_builtin self other
+    != self other = (self == other).not
+    < self other = Comparable.less_than_builtin self other
+    <= self other = Comparable.less_than_builtin self other || Comparable.equals_builtin self other
```
I'm surprised we don't have a `less_than_or_equals` builtin.

radeusgd reviewed (commented):
I see there were many builtin nodes removed for the primitive numeric types.

From the perspective of code clarity - that seems good and I like it.

But do we have benchmarks that show that this will not degrade performance of comparisons on primitive numbers? These should be able to PE into simple instructions on the CPU. If we have the more complex comparison nodes + stuff like `<=` is now two operations linked with a `||`; will it still compile to good code on the fastpath? I'm worried if this will not suffer from the same issue as #6191 - i.e. the nodes for `Integer.<` will be larger than they have to, potentially hurting performance. Did you check that this is not the case?

radeusgd commented on engine/runtime/src/main/java/org/enso/interpreter/node/expression/builtin/ordering/SortVectorNode.java:
```diff
@@ -0,0 +1,836 @@
+    }
+  }
+
+  /**
+   * Comparator for any values. This comparator compares the values by calling back to Enso (by
+   * {@link #compareFunc}), rather than using compare nodes (i.e. {@link LessThanNode}). directly,
+   * as opposed to {@link DefaultSortComparator}.
+   */
+  private final class GenericSortComparator extends SortComparator {
```
nitpick: This file is pretty big. Wouldn't it make it more readable to move this class to a separate file?

Akirathan replied:
I am afraid not. The inner classes are not `static` on purpose - they use the reference to the outer class.

radeusgd replied:
Oh! I missed this fact. Then ignore this.

radeusgd commented on engine/runtime/src/main/java/org/enso/interpreter/node/expression/builtin/ordering/SortVectorNode.java:
```diff
@@ -0,0 +1,836 @@
+    }
+  }
+
+  private int getBuiltinTypeCost(Object builtinType) {
```
What is the meaning of _cost_ in this context? Please document.

radeusgd replied:
Is `Cost` the right name here?

Akirathan replied:
Two values are assigned an integer *cost*, and the one that has lower *cost* is lower. I guess that `weight` would also be acceptable name. Any better name suggestions? If you look into the call site of this method, it is pretty clear what it does and how it should behave.

radeusgd replied:
Yeah. Weight sounds better IMO.
I guess you are right, maybe a note that this weight determines the relative ordering of each of primitive types?

radeusgd commented on engine/runtime/src/main/java/org/enso/interpreter/node/expression/builtin/ordering/SortVectorNode.java:
```diff
@@ -0,0 +1,836 @@
+      this.interop = interop;
+    }
+
+    @TruffleBoundary
+    protected void attachIncomparableValuesWarning(Object x, Object y) {
+      var xStr = toTextNode.execute(x).toString();
+      var yStr = toTextNode.execute(y).toString();
+      String warnText = "Values " + xStr + " and " + yStr + " are incomparable";
+      warnings.add(warnText);
+    }
```
We will be adding a warning for each pair of incomparable values when sorting.

If I sort a vector of 1M of values, each of them being incomparable, how many of these warnings will I get?

I'm thinking that we may want to report only the first 3-10 of these and then report just a counter _'and N more values that were incomparable.'_

See for example `Additional_Warnings` in the Table library which is a similar concept. When doing Table aggregations we report the first 10 issues and then just report the count of additional problems. We don't want 1000+ of the same problem just with different values in it attached, it will not add much more information.

More importantly, sorting 1M elements that are comparable vs incomparable shouldn't have vastly different performance characteristics.

I'm not pushing to necessarily do it _now_, but ideally we should at least create a ticket to ensure that large incomparable vectors are handled well by this. Ideally with adding a test that sorts a vector of at least 10k+ values and measures the amount of warnings, possibly also a benchmark.

Akirathan replied:
Tracked in #6283, as a follow-up.

Commit:
Remove TruffleLogger from SortVectorNode

Commit:
Small review refactorings

radeusgd commented on distribution/lib/Standard/Base/0.0.0-dev/src/Data/Vector.enso:
```diff
@@ -915,11 +923,7 @@ type Vector a
             True -> self.map it-> Comparable.from it
             False -> self.map it-> Comparable.from (on it)
         compare_funcs = comps.map (it-> it.compare)
-        problem_behavior = case on_incomparable of
```
^^^ I cannot comment above due to lack of changes. The changes to `Array` Example with backticks added to `1` and `My_Type` should be ported to the Vector docstring as well.

Akirathan replied:
There is actually a test that checks whether docs on methods from `Vector` are mirrored to `Array`.

radeusgd commented on distribution/lib/Standard/Base/0.0.0-dev/src/Data/Numbers.enso:
```diff
@@ -280,6 +280,8 @@ type Number
 
     ## Checks equality of numbers, using an `epsilon` value.
 
+       Throws `Incomparable_Values` error if `Number.nan` is used as `self` or `that`.
```
```suggestion
       ! Error Conditions
         If either of the arguments is `Number.nan`, an `Incomparable_Values` error is raised.
```


I don't want to impose a style too much, but I think this form may look a bit better.

radeusgd reviewed (approved):
I'll appreciate applying the two suggestions. Other than that - it all looks very good! I'm glad vector sort will be improved now. And I really appreciate very good unit tests.

Akirathan commented:
Scheduled benchmarks in https://github.com/enso-org/enso/actions/runs/4702222179.

I have checked `test/Benchmarks/src/Table/Sorting.enso` benchmark locally, and all the benchmarks have better performance, except for "Table.order_by object", that regressed by 25%. That particular benchmark creates a lot of `My` atoms with a custom `My_Comparator` and are move into a table and `order_by` is called, which calls into Java and then back into Enso (in `ObjectComparator`). There is no simple solution how to make that faster in general, at least not now. I am not sure why there is such regression, though, and I am not going to investigate anymore now.

<blockquote><img src="https://opengraph.githubassets.com/6cf40e2b01a459b39eb61c641c3436d2791fe63511fb25ddbccd7051acb83a1a/enso-org/enso" width="48" align="right"><div><img src="https://github.githubassets.com/favicons/favicon.svg" height="14"> GitHub</div><div><strong><a href="https://github.com/enso-org/enso">Benchmark Engine · enso-org/enso@8268e02</a></strong></div><div>Hybrid visual and textual functional programming. Contribute to enso-org/enso development by creating an account on GitHub.</div></blockquote>

JaroslavTulach commented on distribution/lib/Standard/Base/0.0.0-dev/src/Data/Array.enso:
```diff
@@ -197,8 +222,17 @@ type Array
+
+             [My_Type.Value 'hello', 1].to_array.sort == [1, My_Type.Value 'hello'].to_array
+    sort : Sort_Direction -> (Any -> Any)|Nothing -> (Any -> Any -> (Ordering|Nothing))|Nothing -> Problem_Behavior -> Vector Any ! Incomparable_Values
+    sort self (order = Sort_Direction.Ascending) on=Nothing by=Nothing on_incomparable=Problem_Behavior.Ignore =
```
Is the documentation consistent? I see:

> warning will be attached to the resulting array.

vs.

> `on_incomparable=Problem_Behavior.Ignore`

Akirathan replied:
Fixed

JaroslavTulach commented on distribution/lib/Standard/Base/0.0.0-dev/src/Data/Numbers.enso:
```diff
@@ -457,54 +461,6 @@ type Decimal
     ^ : Number -> Number
     ^ self that = @Builtin_Method "Decimal.^"
 
-    ## Checks if this is greater than that.
```
Wow that's a remove of lot of code. I assume it is a result of https://github.com/enso-org/enso/pull/5998/commits/0df66b10806036594cf51b13df6062ab32680689 and will lead to overall unification of the checks in a single `LessThenNode`.

Akirathan replied:
Unfortunately, I had to revert this once again, because it caused a lot of performance regressions in vector operations with primitive values. Explanation in https://github.com/enso-org/enso/pull/5998#issuecomment-1509947907. We can deal with this later, though.

JaroslavTulach replied:
I see.

JaroslavTulach commented on distribution/lib/Standard/Base/0.0.0-dev/src/Data/Ordering.enso:
````diff
@@ -27,7 +27,7 @@ from project.Data.Boolean import all
     ```
     type Comparator T
-        compare : T -> T -> (Ordering|Nothing)
+        compare : T -> T -> (Ordering|Nothing) ! Incomparable_Values
````
`Nothing` used to be there to signal "incomparable values". Do we now have two ways to signal the same thing? Does returning `Nothing` means the same thing as `Error.throw Incomparable_Values`?

Akirathan replied:
Sorry, this was my mistake. The signature should not change. Rewriting this back to `compare : T -> T -> (Ordering|Nothing)`.

JaroslavTulach commented on distribution/lib/Standard/Base/0.0.0-dev/src/Data/Ordering.enso:
```diff
@@ -170,7 +170,7 @@ type Ordering
 
     ## Compares to values and returns an Ordering
-    compare : Any -> Any -> Ordering ! (Incomparable_Values | Type_Error)
+    compare : Any -> Any -> Ordering ! Incomparable_Values
```
This is a simplification of the API - good for users of `Ordering.compare` that they don't need to think about two different errors.

Akirathan replied:
This was possible because https://github.com/enso-org/enso/pull/5998/commits/603ab95a184871a49ac8b21fe5f8bd4122647aca - I decided that if comparing two values with different comparators, `Incomparable_Values` is thrown rather than `Type_Error`. I think this is, indeed, a change for the better.

JaroslavTulach commented on engine/runtime/src/test/java/org/enso/interpreter/test/DebuggingEnsoTest.java:
```diff
@@ -113,6 +113,8 @@ private Value createEnsoMethod(String source, String methodName) {
   public void recursiveFactorialCall() {
     final Value facFn = createEnsoMethod("""
+    from Standard.Base.Data.Ordering import all
+    
```
Without this import we cannot use `<=` anymore.
- result of 0df66b108060
- also reported as #5610

JaroslavTulach commented on test/Tests/src/Data/Vector_Spec.enso:
```diff
@@ -557,11 +557,8 @@ type_spec name alter = Test.group name <|
-        alter [Date_Time.new 1999 1 1 12 30, Date.new 1999] . sort . should_fail_with Incomparable_Values
-        alter [Date_Time.new 1999 1 1 12 30, Time_Of_Day.new 12 30] . sort . should_fail_with Incomparable_Values
-        Test.expect_panic_with ([3,2,1].to_array.sort_builtin 42) Type_Error
+        ["aa", 2].sort . should_equal [2, "aa"]
```
Please use the `alter` prefix to make sure the test runs with `Vector` as well as with `Array` - e.g.:
```
alter ["aa", 2] . sort . should_equal [2, "aa"]
```
please use this `alter` function everywhere in this `_Spec`.

Akirathan replied:
OK, done

JaroslavTulach commented on engine/runtime/src/main/java/org/enso/interpreter/node/expression/builtin/ordering/SortArrayNode.java:
```diff
@@ -0,0 +1,29 @@
+import org.enso.interpreter.runtime.state.State;
+
+/** Just a wrapper for {@code Array.sort_builtin} that delegates to {@code Vector.sort_builtin}. */
+@BuiltinMethod(type = "Array", name = "sort_builtin")
```
After #6218 there remained only one difference between `Vector` and `Array` - `Array` has additional `sort_buitin`. `Vector` sorting used to be done by copying the `Vector` content into `Array`, sorting that one in place and wrapping it into `Vector`.

`Array.sort_buitin` was (supposed to be) a private method. Implementation of `Array.sort` delegated to `Vector.sort` and returned new copy of the original array.

How that is planned to be done now? Can we have just a single `sort_buitin`? Do we need both for some reason?

Akirathan replied:
The only reason why `Array.sort_builtin` exist now is because #6282. When #6282 is fixed, it can be removed. I will put a link to this comment in that issue, so that we know that we should remove this `SortArrayNode` altogether once that issue is resolved.

JaroslavTulach commented on engine/runtime/src/main/java/org/enso/interpreter/node/expression/builtin/ordering/SortVectorNode.java:
```diff
@@ -0,0 +1,800 @@
+ * Sorts a vector with elements that have only Default_Comparator, thus, only elements with a
+ * builtin type, which is the most common scenario for sorting.
+ *
+ * <p>TODO: Max number of attached Incomparable values warnings? - hardcode or pass via a new
```
Hardcode to five "types" of warnings, in my opinion. A "type" of warning is a clash between two groups of comparable items. The last warning could say _"there is N more incomparable problems"_.

Akirathan replied:
I will remove this TODO comment altogether. Treatment of warnings by this node is tracked in #6283.

JaroslavTulach commented on engine/runtime/src/main/java/org/enso/interpreter/node/expression/builtin/ordering/SortVectorNode.java:
```diff
@@ -0,0 +1,800 @@
+      elems = new Object[(int) size];
+      for (int i = 0; i < size; i++) {
+        if (interop.isArrayElementReadable(self, i)) {
+          elems[i] = hostValueToEnsoNode.execute(interop.readArrayElement(self, i));
```
Use of `hostValueToEnsoNode` is good in the context of https://github.com/enso-org/enso/pull/6280#discussion_r1167372095

Akirathan replied:
Right. I wrapped all the other element reads via `InteropLibrary` with `HostValueToEnsoNode`.

JaroslavTulach commented on engine/runtime/src/main/java/org/enso/interpreter/node/expression/builtin/ordering/SortVectorNode.java:
```diff
@@ -0,0 +1,800 @@
+      int size = (int) interop.getArraySize(vector);
+      List<T> res = new ArrayList<>(size);
+      for (int i = 0; i < size; i++) {
+        Object elem = interop.readArrayElement(vector, i);
```
Possibly missing conversion to Enso values (https://github.com/enso-org/enso/pull/6280#discussion_r1167372095) - maybe we should have a generic `AtNode` to read an Enso value from anything array-like.

Akirathan replied:
Done. See https://github.com/enso-org/enso/pull/5998#discussion_r1167634829

JaroslavTulach reviewed (approved):
- Use `alter` in the `Vector_Spec.enso`
- Don't forget to convert interop values with `HostToEnsoValueNode`

Commit:
Revert "Remove the remaining comparison operator overrides for numbers."

This reverts commit 0df66b10806036594cf51b13df6062ab32680689.

Akirathan commented:
## Benchmark results
Baseline from develop:
- https://github.com/enso-org/enso/actions/runs/4684019136
- On commit https://github.com/enso-org/enso/commit/4e92d766653623d15fc693b2ee575bf676fcd8ac

I had to revert https://github.com/enso-org/enso/pull/5998/commits/0df66b10806036594cf51b13df6062ab32680689 ("Remove the remaining comparison operator overrides for numbers"), as that caused a huge performance regressions on benchmarks dealing with vectors with primitive values. This commit was not essential for this PR, and we can deal with it later.

After the revert, I tried running and comparing some benchmarks locally. Comparison of my recent version (https://github.com/enso-org/enso/pull/5998/commits/254d57982dcab53a1717238a408a40e8855bd087) with the baseline is:
- `test/Benchmarks/src/Table/Sorting.enso`:
  - Table.order_by ints: -5% (means -5% milliseconds, so it is better 5% better than before)
  - Vector.sort ints: -5%
  - Table.order_by_dates: -7%
  - **Vector.sort dates -36%**
  - Table.order_by objects +1%
  - **Vector.sort objects -64%**
- Benchmarks from `org.enso.interpreter.bench.benchmarks.semantic.ArrayProxyBenchmarks` stayed roughly the same, with one exception of:
  - `sumOverVector` that got +40%. I have no idea why, the changes in this PR should not affect that benchmark. Tried to look into the Truffle compilation logs and did not discovered anything interesting. I will let this one go now.

## Conclusion
Overall, the benchmark results seem pretty good. Sorting of vector with primitive values is roughly the same, but we can see huge improvements for sorting of vector with some atoms (-64%) and dates (-36%). Will merge this PR ASAP.

<blockquote><img src="https://opengraph.githubassets.com/6cf40e2b01a459b39eb61c641c3436d2791fe63511fb25ddbccd7051acb83a1a/enso-org/enso" width="48" align="right"><div><img src="https://github.githubassets.com/favicons/favicon.svg" height="14"> GitHub</div><div><strong><a href="https://github.com/enso-org/enso">Benchmark Engine · enso-org/enso@4e92d76</a></strong></div><div>Hybrid visual and textual functional programming. Contribute to enso-org/enso development by creating an account on GitHub.</div></blockquote>

Commit:
Improve bench_download.py tool's `--compare` functionality.

- Output table is sorted by benchmark labels.
- Do not fail when there are different benchmark labels in both runs.

Commit:
Wrap potential interop values with `HostValueToEnsoNode`

Commit:
Use alter function in Vector_Spec

Commit:
Update docs

Commit:
Merge branch 'develop' into wip/akirathan/sort-incomparable-5742

# Conflicts:
#	test/micro-distribution/lib/Standard/Base/0.0.0-dev/src/Any.enso

Commit:
Invalid comparison throws Incomparable_Values rather than Type_Error

Commit:
Number comparison builtin methods return Nothing in case of incomparables
